api/v1beta1/identity_types.go

```diff
@@ -18,6 +18,7 @@ package v1beta1
 
 // OpenStackIdentityReference is a reference to an infrastructure
 // provider identity to be used to provision cluster resources.
+// +kubebuilder:validation:XValidation:rule="(!has(self.region) && !has(oldSelf.region)) || self.region == oldSelf.region",message="region is immutable"
```

Comment:
yes, `"self == oldSelf"` as you proposed, permit to set a value on an existing machine without region field defined. To do that I had to create rule on top level object because `has(self)` seems not valid syntax.

Reply:
Thanks. Turns out I was missing some tests in the image API and this was broken: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/2197

Reply:
Glad to have been helpful !